File: libs/statemanager/src/main/kotlin/mono/state/command/mouse/MoveShapeMouseCommand.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state.command.mouse

import mono.common.MouseCursor
import mono.graphics.geo.MousePointer
import mono.graphics.geo.Point
import mono.shape.shape.AbstractShape
import mono.state.command.CommandEnvironment
import mono.state.command.mouse.MouseCommand.CommandResultType
import mono.state.utils.UpdateShapeBoundHelper

/**
 * A [MouseCommand] for moving selected shapes.
 */
internal class MoveShapeMouseCommand(
    private val shapes: Set<AbstractShape>,
    relatedShapes: Sequence<AbstractShape>
) : MouseCommand {
    override val mouseCursor: MouseCursor = MouseCursor.MOVE

    private val initialPositions = shapes.associate { it.id to it.bound.position } +
        relatedShapes.associate { it.id to it.bound.position }

    override fun execute(
        environment: CommandEnvironment,
        mousePointer: MousePointer
    ): CommandResultType {
        val offset = when (mousePointer) {
            is MousePointer.Drag -> mousePointer.boardCoordinate - mousePointer.mouseDownPoint
            is MousePointer.Up -> mousePointer.boardCoordinate - mousePointer.mouseDownPoint
            is MousePointer.Down,
            is MousePointer.Click,
            is MousePointer.DoubleClick,
            is MousePointer.Move,
            MousePointer.Idle -> Point.ZERO
        }

        UpdateShapeBoundHelper.moveShapes(
            environment,
            shapes,
            isUpdateConfirmed = mousePointer is MousePointer.Up
        ) { shape -> initialPositions[shape.id]?.let { it + offset } }

        environment.updateInteractionBounds()

        val isDone = mousePointer is MousePointer.Up || mousePointer == MousePointer.Idle
        return if (isDone) CommandResultType.DONE else CommandResultType.WORKING
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/command/mouse/move-shape-mouse-command.ts
import { Point } from "$libs/graphics-geo/point";
import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import type { CommandEnvironment } from "$mono/state-manager/command-environment";
import { MouseCommand } from "$mono/state-manager/command/mouse/mouse-command";
import { moveShapes } from "$mono/state-manager/utils/UpdateShapeBoundHelper";
import { MouseCursor } from "$mono/workspace/mouse/cursor-type";
import { type MousePointer, MousePointerType } from "$mono/workspace/mouse/mouse-pointer";

/**
 * A [MouseCommand] for moving selected shapes.
 */
export class MoveShapeMouseCommand implements MouseCommand {
    mouseCursor: MouseCursor = MouseCursor.MOVE;

    private readonly initialPositions: Map<string, Point> = new Map();

    constructor(
        private shapes: AbstractShape[],
        relatedShapes: AbstractShape[]
    ) {
        // Add positions from shapes
        for (const shape of shapes) {
            this.initialPositions.set(shape.id, shape.bound.position);
        }

        // Add positions from relatedShapes
        for (const shape of relatedShapes) {
            this.initialPositions.set(shape.id, shape.bound.position);
        }
    }

    execute(environment: CommandEnvironment, mousePointer: MousePointer): MouseCommand.CommandResultType {
        const offset = this.getOffset(mousePointer);
        const mouseType = mousePointer.type;

        moveShapes(environment, this.shapes, mouseType === MousePointerType.UP, shape => {
            const initialPosition = this.initialPositions.get(shape.id);
            return initialPosition ? initialPosition.plus(offset) : null;
        });
        environment.updateInteractionBounds();

        const isDone = mouseType === MousePointerType.UP || mouseType === MousePointerType.IDLE;
        return isDone ? MouseCommand.CommandResultType.DONE : MouseCommand.CommandResultType.WORKING;
    }

    private getOffset(mousePointer: MousePointer): Point {
        switch (mousePointer.type) {
            case MousePointerType.UP:
                return mousePointer.boardCoordinate.minus(mousePointer.mouseDownBoardCoordinate);
            case MousePointerType.DRAG:
                return mousePointer.boardCoordinate.minus(mousePointer.mouseDownBoardCoordinate);
            case MousePointerType.IDLE:
            case MousePointerType.MOVE:
            case MousePointerType.DOWN:
            case MousePointerType.CLICK:
            case MousePointerType.DOUBLE_CLICK:
                return Point.ZERO;
        }
    }
}
